File: 2016/day1/pb.lua
local x, y = 0, 0
local facing = 0

local result = {}
local visited = {}

for line in io.lines("./input.txt") do
	for value in string.gmatch(line, "[^,]+") do
		value = value:match("^%s*(.-)%s*$")
		table.insert(result, value)
	end
end

local first_repeat = nil

local function visit(x, y)
	local key = x .. "," .. y
	if visited[key] then
		return key
	else
		visited[key] = true
		return nil
	end
end

for _, v in ipairs(result) do
	local turn, steps = v:match("^(%a)(%d+)$")
	steps = tonumber(steps)

	if turn == "R" then
		facing = (facing + 1) % 4
	elseif turn == "L" then
		facing = (facing + 3) % 4
	else
		error("Invalid turn: " .. turn)
	end

	for i = 1, steps do
		if facing == 0 then
			y = y + 1
		elseif facing == 1 then
			x = x + 1
		elseif facing == 2 then
			y = y - 1
		elseif facing == 3 then
			x = x - 1
		end

		if not first_repeat then
			local repeat_key = visit(x, y)
			if repeat_key then
				first_repeat = { x = x, y = y }
			end
		end
	end
end

if first_repeat then
	print(first_repeat.x .. "," .. first_repeat.y)
	print(math.abs(first_repeat.x) + math.abs(first_repeat.y))
else
	print("No repeated coordinate found")
end

File: 2016-py/day1/pb.py
def visit(x, y, visited):
    key = (x, y)

    if key in visited:
        return True

    visited.add(key)
    return False


def main():
    x, y = 0, 0
    facing = 0  # north=0, east=1, south=2, west=3

    result = []
    visited = {(0, 0)}

    with open("input.txt", "r") as fin:
        for line in fin:
            values = line.strip().split(",")
            for value in values:
                result.append(value.strip())

    first_repeat = None

    for instruction in result:
        turn = instruction[0]
        steps = int(instruction[1:])

        if turn == "R":
            facing = (facing + 1) % 4
        elif turn == "L":
            facing = (facing + 3) % 4
        else:
            raise ValueError(f"Invalid turn: {turn}")

        for _ in range(steps):
            if facing == 0:
                y += 1
            elif facing == 1:
                x += 1
            elif facing == 2:
                y -= 1
            elif facing == 3:
                x -= 1

            if first_repeat is None and visit(x, y, visited):
                first_repeat = (x, y)

    if first_repeat is not None:
        print(f"{first_repeat[0]},{first_repeat[1]}")
        print(abs(first_repeat[0]) + abs(first_repeat[1]))
    else:
        print("No repeated coordinate found")


if __name__ == "__main__":
    main()

